server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.util.timer;
+
+import org.apache.kafka.common.utils.KafkaThread;
+import org.apache.kafka.common.utils.Time;
+
+import java.util.concurrent.DelayQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class SystemTimer implements Timer {
+    // timeout timer
+    private final ExecutorService taskExecutor;
+    private final DelayQueue<TimerTaskList> delayQueue;
+    private final AtomicInteger taskCounter;
+    private final TimingWheel timingWheel;
+
+    // Locks used to protect data structures while ticking
+    private final ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    private final ReentrantReadWriteLock.ReadLock readLock = readWriteLock.readLock();
+    private final ReentrantReadWriteLock.WriteLock writeLock = readWriteLock.writeLock();
+
+    public SystemTimer(String executorName) {
+        this(executorName, 1, 20, Time.SYSTEM.hiResClockMs());
+    }
+
+    public SystemTimer(
+        String executorName,
+        long tickMs,
+        int wheelSize,
+        long startMs
+    ) {
+        this.taskExecutor = Executors.newFixedThreadPool(1,
+            runnable -> KafkaThread.nonDaemon("executor-" + executorName, runnable));
+        this.delayQueue = new DelayQueue<>();
+        this.taskCounter = new AtomicInteger(0);
+        this.timingWheel = new TimingWheel(
+            tickMs,
+            wheelSize,
+            startMs,
+            taskCounter,
+            delayQueue
+        );
+    }
+
+    public void add(TimerTask timerTask) {
+        readLock.lock();
```

Comment:
we are using this pattern of locking very often now (also used in TS related code for leader epoch snapshot). Could we perhaps extract this out into a utility which takes incoming lambda and executes it in a lock, similar to https://github.com/apache/kafka/blob/513e1c641d63c5e15144f9fcdafa1b56c5e5ba09/core/src/main/scala/kafka/utils/CoreUtils.scala#L182

Reply:
Personally, I am not a big fan of those helper methods for locks in our Java code. I think that we could consider this separately.

Reply:
ok